base/complex.jl

```diff
@@ -1019,3 +1019,11 @@ function complex(A::AbstractArray{T}) where T
     end
     convert(AbstractArray{typeof(complex(zero(T)))}, A)
 end
+
+function complex(A::AbstractArray{T1}, B::AbstractArray{T2}) where {T1, T2}
+    if !isconcretetype(T1) || !isconcretetype(T2)
+        error("`complex` not defined on abstractly-typed arrays; please convert to a more specific type")
```

Comment:
???

Reply:
I copied the error from the existing [one-array method](https://github.com/JuliaLang/julia/blob/1c05666aacc4d7bc77e91987278947f06ba013b6/base/complex.jl#L1018). I suppose someone might want to construct complex arrays with `missing`s, or from some generically-typed array, but that hasn't been previously allowed:
```
julia> complex(Array{Real,1}())
ERROR: `complex` not defined on abstractly-typed arrays; please convert to a more specific type
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] complex(::Array{Real,1}) at ./complex.jl:1013
 [3] top-level scope at none:0
```

Reply:
Maybe `!isconcretetype(promote_type(T1, T2))` is better?

Reply:
I think we should get rid of the concrete type restriction from the one-array method as well.  It probably should be:
```jl
complex(A::AbstractArray) = map(complex, A)
complex(A::AbstractArray{<:Real}) = A
complex(R::AbstractArray{<:Real,N}, I::AbstractArray{<:Real,N}) where {N} = map(complex, R, I)
```
and let `map` take care of figuring out the element type.  Note that `map` will also verify that `R` and `I` have the same shape, and the type signature ensures that the dimensionality is the same.

Reply:
The above definition also ensures that if we ever update the scalar `complex` function to work with missing data, then the array function will automatically be upgraded as well.